File: Moodist/SoundMixer.swift
import AVFoundation
import Combine
import SwiftUI

/// Plays any number of looping sounds at once, each with its own volume.
///
/// Selecting a sound while the mixer is playing starts that loop immediately, so the
/// mix changes live rather than on the next play — matching the web app's behaviour.
@MainActor
final class SoundMixer: ObservableObject {
    /// Volume per selected sound id. Membership here *is* selection.
    @Published private(set) var volumes: [String: Double] = [:]
    @Published private(set) var isPlaying = false

    /// An ordered list of presets to step through with next/previous. Empty when the
    /// current mix wasn't started from "Play All" — the skip controls are inert then.
    @Published private(set) var queue: [Preset] = []
    /// Index into `queue` of the preset currently playing, or nil when there's no queue.
    @Published private(set) var queueIndex: Int?

    /// The name of the preset the current mix was loaded from, so the player shows the
    /// preset's title. Cleared the moment the user hand-edits the selection.
    @Published private(set) var currentPresetName: String?

    /// Scales every sound in the mix. Each player's level is master × its own volume.
    @Published var masterVolume: Double = 0.8 {
        didSet { applyMasterVolume() }
    }

    private var players: [String: AVAudioPlayer] = [:]
    private let nowPlaying = NowPlayingController()

    /// Set when the system interrupts us (call, Siri) so we know whether an
    /// auto-resume is ours to make.
    private var wasPlayingBeforeInterruption = false

    private static let fadeDuration: TimeInterval = 0.25
    private static let defaultVolume: Double = 0.5

    init() {
        configureSession()

        nowPlaying.onPlay = { [weak self] in self?.resumeAll() }
        nowPlaying.onPause = { [weak self] in self?.pauseAll() }
        observeInterruptions()
        refreshNowPlaying()
    }

    // MARK: - Queries

    var selectedIDs: [String] { Array(volumes.keys) }
    var selectedCount: Int { volumes.count }
    var hasSelection: Bool { !volumes.isEmpty }

    /// Whether a "Play All" queue is currently steering the mix.
    var hasQueue: Bool { queueIndex != nil && queue.count > 1 }

    func isSelected(_ sound: Sound) -> Bool { volumes[sound.id] != nil }

    func volume(for sound: Sound) -> Double { volumes[sound.id] ?? Self.defaultVolume }

    /// The sounds currently in the mix, in catalog order, for the player bar's label.
    var selectedSounds: [Sound] {
        SoundCatalog.allSounds.filter { volumes[$0.id] != nil }
    }

    /// Artwork follows the first sound's category, so the player reflects the mix.
    var artworkName: String? { selectedSounds.first?.categoryID }

    var mixTitle: String {
        if let name = currentPresetName { return name }
        let sounds = selectedSounds
        guard let first = sounds.first else { return "Nothing playing" }
        return sounds.count == 1 ? first.label : "\(first.label) + \(sounds.count - 1) more"
    }

    var mixSubtitle: String {
        guard hasSelection else { return "No sounds selected" }
        guard isPlaying else { return "Paused" }
        return "\(selectedCount) sound\(selectedCount == 1 ? "" : "s") playing"
    }

    // MARK: - Selection

    func toggle(_ sound: Sound) {
        if isSelected(sound) {
            deselect(sound)
        } else {
            select(sound)
        }
    }

    func select(_ sound: Sound) {
        guard volumes[sound.id] == nil else { return }
        // Adding a sound by hand means this is no longer the saved preset.
        currentPresetName = nil
        volumes[sound.id] = Self.defaultVolume

        // First sound picked from a stopped mixer starts playback; otherwise the new
        // loop joins the mix already in progress.
        if !isPlaying {
            isPlaying = true
            startAllSelected()
        } else {
            start(sound)
        }
        refreshNowPlaying()
    }

    func deselect(_ sound: Sound) {
        // Removing a sound by hand means this is no longer the saved preset.
        currentPresetName = nil
        volumes[sound.id] = nil
        fadeOutAndStop(id: sound.id)

        if volumes.isEmpty {
            isPlaying = false
        }
        refreshNowPlaying()
    }

    func setVolume(_ value: Double, for sound: Sound) {
        guard volumes[sound.id] != nil else { return }
        volumes[sound.id] = value
        players[sound.id]?.setVolume(level(for: sound.id), fadeDuration: 0)
    }

    func clearAll() {
        clearQueue()
        currentPresetName = nil
        for id in players.keys { fadeOutAndStop(id: id) }
        volumes.removeAll()
        isPlaying = false
        refreshNowPlaying()
    }

    /// Replaces the mix with four random sounds at random volumes and starts playing —
    /// mirrors the web app's shuffle.
    func shuffle() {
        clearQueue()
        currentPresetName = nil
        for id in players.keys { fadeOutAndStop(id: id) }

        let picked = SoundCatalog.allSounds.shuffled().prefix(4)
        guard !picked.isEmpty else {
            volumes.removeAll()
            isPlaying = false
            refreshNowPlaying()
            return
        }

        var next: [String: Double] = [:]
        for sound in picked {
            next[sound.id] = Double.random(in: 0.2...1)
        }
        volumes = next
        isPlaying = true
        startAllSelected()
        refreshNowPlaying()
    }

    // MARK: - Presets

    /// Captures the current mix so it can be restored later.
    func snapshot(named name: String) -> Preset {
        Preset(name: name, volumes: volumes, masterVolume: masterVolume)
    }

    /// Replaces the current mix with a saved preset and starts playing it. Tapping a
    /// preset directly ends any active queue — the user has stepped off the playlist.
    func restore(_ preset: Preset) {
        clearQueue()
        loadMix(from: preset)
    }

    // MARK: - Queue

    /// Starts a playlist: loads the first preset (or the one at `index`) and keeps the
    /// rest queued so next/previous can step through them.
    func playQueue(_ presets: [Preset], startAt index: Int = 0) {
        let ordered = presets.filter { !$0.volumes.isEmpty }
        guard !ordered.isEmpty else { return }
        queue = ordered
        playQueueItem(at: min(max(index, 0), ordered.count - 1))
    }

    /// Advances to the next preset in the queue, wrapping to the start at the end.
    func next() {
        guard let i = queueIndex, !queue.isEmpty else { return }
        playQueueItem(at: (i + 1) % queue.count)
    }

    /// Steps back to the previous preset in the queue, wrapping to the end at the start.
    func previous() {
        guard let i = queueIndex, !queue.isEmpty else { return }
        playQueueItem(at: (i - 1 + queue.count) % queue.count)
    }

    private func playQueueItem(at index: Int) {
        guard queue.indices.contains(index) else { return }
        queueIndex = index
        loadMix(from: queue[index])
    }

    private func clearQueue() {
        queue = []
        queueIndex = nil
    }

    /// Drops the "playing from a preset" label so the mix is titled by its sounds again.
    /// Used for shared mixes, which aren't saved presets.
    func clearCurrentPresetName() {
        currentPresetName = nil
        refreshNowPlaying()
    }

    /// The audio work behind restoring a preset, shared by direct restore and the queue.
    private func loadMix(from preset: Preset) {
        for id in players.keys { fadeOutAndStop(id: id) }
        currentPresetName = preset.name
        volumes = preset.volumes.filter { SoundCatalog.sound(id: $0.key) != nil }
        masterVolume = preset.masterVolume

        guard !volumes.isEmpty else {
            currentPresetName = nil
            isPlaying = false
            refreshNowPlaying()
            return
        }
        isPlaying = true
        startAllSelected()
        refreshNowPlaying()
    }

    private func refreshNowPlaying() {
        nowPlaying.update(
            title: mixTitle,
            subtitle: hasSelection ? mixSubtitle : "",
            artworkName: artworkName,
            isPlaying: isPlaying,
            hasMix: hasSelection
        )
    }

    private func level(for id: String) -> Float {
        Float((volumes[id] ?? Self.defaultVolume) * masterVolume)
    }

    private func applyMasterVolume() {
        guard isPlaying else { return }
        for (id, player) in players {
            player.setVolume(level(for: id), fadeDuration: 0)
        }
    }

    // MARK: - Transport

    func togglePlayPause() {
        guard hasSelection else { return }
        isPlaying ? pauseAll() : resumeAll()
    }

    /// Also the target for the lock screen / Control Centre play button.
    private func resumeAll() {
        guard hasSelection else { return }
#if os(iOS)
        // The session may have been deactivated by an interruption.
        try? AVAudioSession.sharedInstance().setActive(true)
#endif
        isPlaying = true
        startAllSelected()
        refreshNowPlaying()
    }

    private func pauseAll() {
        isPlaying = false
        refreshNowPlaying()
        for (id, player) in players {
            player.setVolume(0, fadeDuration: Self.fadeDuration)
            DispatchQueue.main.asyncAfter(deadline: .now() + Self.fadeDuration) { [weak self] in
                // A resume may have landed during the fade; leave it playing if so.
                guard let self, !self.isPlaying else { return }
                player.pause()
                player.volume = self.level(for: id)
            }
        }
    }

    private func startAllSelected() {
        for sound in selectedSounds { start(sound) }
    }

    // MARK: - Players

    private func start(_ sound: Sound) {
        guard let player = player(for: sound) else { return }

        player.numberOfLoops = -1
        if !player.isPlaying {
            player.volume = 0
            player.play()
        }
        player.setVolume(level(for: sound.id), fadeDuration: Self.fadeDuration)
    }

    private func fadeOutAndStop(id: String) {
        guard let player = players[id] else { return }
        player.setVolume(0, fadeDuration: Self.fadeDuration)
        DispatchQueue.main.asyncAfter(deadline: .now() + Self.fadeDuration) { [weak self] in
            player.stop()
            player.currentTime = 0
            self?.players[id] = nil
        }
    }

    private func player(for sound: Sound) -> AVAudioPlayer? {
        if let existing = players[sound.id] { return existing }
        guard let url = sound.url else {
            assertionFailure("Missing bundled audio for \(sound.id)")
            return nil
        }
        guard let player = try? AVAudioPlayer(contentsOf: url) else { return nil }
        player.numberOfLoops = -1
        player.prepareToPlay()
        players[sound.id] = player
        return player
    }

    private func configureSession() {
#if os(iOS)
        let session = AVAudioSession.sharedInstance()
        // .playback keeps the mix going with the app backgrounded or the screen
        // locked, and routes the hardware volume buttons to it.
        try? session.setCategory(.playback, mode: .default)
        try? session.setActive(true)
#endif
        // macOS needs no audio session; AVAudioPlayer just plays.
    }

    /// Phone calls, Siri and other apps' audio interrupt us; resume only if the
    /// system says we may and we were playing when it started.
    private func observeInterruptions() {
#if os(iOS)
        NotificationCenter.default.addObserver(
            forName: AVAudioSession.interruptionNotification,
            object: AVAudioSession.sharedInstance(),
            queue: .main
        ) { [weak self] note in
            guard let raw = note.userInfo?[AVAudioSessionInterruptionTypeKey] as? UInt,
                  let type = AVAudioSession.InterruptionType(rawValue: raw)
            else { return }

            MainActor.assumeIsolated {
                guard let self else { return }
                switch type {
                case .began:
                    self.wasPlayingBeforeInterruption = self.isPlaying
                    if self.isPlaying { self.pauseAll() }
                case .ended:
                    let options = (note.userInfo?[AVAudioSessionInterruptionOptionKey] as? UInt)
                        .map(AVAudioSession.InterruptionOptions.init(rawValue:)) ?? []
                    if options.contains(.shouldResume), self.wasPlayingBeforeInterruption {
                        self.resumeAll()
                    }
                    self.wasPlayingBeforeInterruption = false
                @unknown default:
                    break
                }
            }
        }
#endif
        // macOS delivers no AVAudioSession interruptions; nothing to observe.
    }
}

File: Yumibun_Android/app/src/main/java/com/toepper/rocks/yumibun/playback/SoundMixer.kt
package com.toepper.rocks.yumibun.playback

import android.content.Context
import android.content.Intent
import android.media.AudioAttributes
import android.media.AudioFocusRequest
import android.media.AudioManager
import android.media.MediaPlayer
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableFloatStateOf
import androidx.compose.runtime.mutableStateListOf
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.runtime.snapshots.SnapshotStateMap
import androidx.core.content.ContextCompat
import com.toepper.rocks.yumibun.data.Preset
import com.toepper.rocks.yumibun.data.Sound
import com.toepper.rocks.yumibun.data.SoundCatalog
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * The application-scoped audio engine: plays any number of looping sounds at once, each
 * with its own volume. Lives beyond the UI so playback continues in the background,
 * driven by [PlaybackService]. Exposes Compose snapshot state for the UI to observe.
 */
class SoundMixer(context: Context) {

    private val appContext = context.applicationContext
    private val prefs = appContext.getSharedPreferences("yumibun", Context.MODE_PRIVATE)
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    /** Volume per selected sound id. Membership here *is* selection. */
    val volumes: SnapshotStateMap<String, Float> = mutableStateMapOf()

    var isPlaying by mutableStateOf(false)
        private set
    var masterVolume by mutableFloatStateOf(prefs.getFloat(KEY_MASTER_VOLUME, 0.8f))
        private set
    var currentPresetName by mutableStateOf<String?>(null)
        private set

    /** An ordered list of presets to step through with next/previous (from "Play All"). */
    val queue = mutableStateListOf<Preset>()
    var queueIndex by mutableStateOf<Int?>(null)
        private set

    /** Whether a "Play All" queue is currently steering the mix. */
    val hasQueue: Boolean get() = queueIndex != null && queue.size > 1

    private val players = HashMap<String, MediaPlayer>()
    private val fadeJobs = HashMap<String, Job>()
    private val currentVolume = HashMap<String, Float>()

    /** Set by [PlaybackService] so state changes refresh the notification. */
    var onStateChanged: (() -> Unit)? = null

    private val audioAttributes = AudioAttributes.Builder()
        .setUsage(AudioAttributes.USAGE_MEDIA)
        .setContentType(AudioAttributes.CONTENT_TYPE_MUSIC)
        .build()

    private val audioManager = appContext.getSystemService(Context.AUDIO_SERVICE) as AudioManager
    private var focusRequest: AudioFocusRequest? = null
    private var pausedForFocusLoss = false

    // MARK: Queries ----------------------------------------------------------

    val selectedCount: Int get() = volumes.size
    val hasSelection: Boolean get() = volumes.isNotEmpty()

    fun isSelected(sound: Sound): Boolean = volumes.containsKey(sound.id)
    fun volumeFor(sound: Sound): Float = volumes[sound.id] ?: DEFAULT_VOLUME

    val selectedSounds: List<Sound>
        get() = SoundCatalog.allSounds.filter { volumes.containsKey(it.id) }

    val artworkName: String?
        get() = selectedSounds.firstOrNull()?.categoryId

    val mixTitle: String
        get() {
            currentPresetName?.let { return it }
            val sounds = selectedSounds
            val first = sounds.firstOrNull() ?: return "Nothing playing"
            return if (sounds.size == 1) first.label else "${first.label} + ${sounds.size - 1} more"
        }

    val mixSubtitle: String
        get() = when {
            !hasSelection -> "No sounds selected"
            !isPlaying -> "Paused"
            else -> "$selectedCount sound${if (selectedCount == 1) "" else "s"} playing"
        }

    fun suggestedPresetName(): String {
        val first = selectedSounds.firstOrNull()?.label ?: "Mix"
        return if (selectedCount <= 1) first else "$first + ${selectedCount - 1}"
    }

    /** Drops the "playing from a preset" label so the mix is titled by its sounds again. */
    fun clearCurrentPresetName() {
        currentPresetName = null
        notifyChanged()
    }

    // MARK: Selection --------------------------------------------------------

    fun toggle(sound: Sound) {
        if (isSelected(sound)) deselect(sound) else select(sound)
    }

    fun select(sound: Sound) {
        if (volumes.containsKey(sound.id)) return
        currentPresetName = null
        volumes[sound.id] = DEFAULT_VOLUME
        if (!isPlaying) {
            isPlaying = true
            requestFocus()
            selectedSounds.forEach { start(it) }
        } else {
            start(sound)
        }
        onPlaybackBegan()
    }

    fun deselect(sound: Sound) {
        currentPresetName = null
        volumes.remove(sound.id)
        fadeOutAndStop(sound.id)
        if (volumes.isEmpty()) {
            isPlaying = false
            stop()
        } else {
            notifyChanged()
        }
    }

    fun setVolume(value: Float, sound: Sound) {
        if (!volumes.containsKey(sound.id)) return
        volumes[sound.id] = value
        players[sound.id]?.let { setLevel(it, sound.id) }
    }

    fun updateMasterVolume(value: Float) {
        masterVolume = value
        prefs.edit().putFloat(KEY_MASTER_VOLUME, value).apply()
        if (isPlaying) players.forEach { (id, p) -> setLevel(p, id) }
    }

    fun clearAll() {
        clearQueue()
        currentPresetName = null
        players.keys.toList().forEach { fadeOutAndStop(it) }
        volumes.clear()
        isPlaying = false
        stop()
    }

    fun shuffle() {
        clearQueue()
        currentPresetName = null
        players.keys.toList().forEach { fadeOutAndStop(it) }
        volumes.clear()
        val picked = SoundCatalog.allSounds.shuffled().take(4)
        picked.forEach { volumes[it.id] = 0.2f + Math.random().toFloat() * 0.8f }
        isPlaying = picked.isNotEmpty()
        if (isPlaying) {
            requestFocus()
            selectedSounds.forEach { start(it) }
            onPlaybackBegan()
        } else {
            stop()
        }
    }

    /** Replaces the current mix with a saved preset. Tapping a preset directly ends any
     *  active queue — the user has stepped off the playlist. */
    fun restore(preset: Preset) {
        clearQueue()
        loadMix(preset)
    }

    private fun loadMix(preset: Preset) {
        players.keys.toList().forEach { fadeOutAndStop(it) }
        volumes.clear()
        preset.volumes.forEach { (id, v) -> if (SoundCatalog.sound(id) != null) volumes[id] = v }
        updateMasterVolume(preset.masterVolume)
        if (volumes.isEmpty()) {
            currentPresetName = null
            isPlaying = false
            stop()
            return
        }
        currentPresetName = preset.name
        isPlaying = true
        requestFocus()
        selectedSounds.forEach { start(it) }
        onPlaybackBegan()
    }

    // MARK: Queue ("Play All") ----------------------------------------------

    /** Starts a playlist: loads the first preset (or the one at [index]) and keeps the
     *  rest queued so next/previous can step through them. */
    fun playQueue(presets: List<Preset>, index: Int = 0) {
        val ordered = presets.filter { it.volumes.isNotEmpty() }
        if (ordered.isEmpty()) return
        queue.clear()
        queue.addAll(ordered)
        playQueueItem(index.coerceIn(0, ordered.size - 1))
    }

    fun next() {
        val i = queueIndex ?: return
        if (queue.isNotEmpty()) playQueueItem((i + 1) % queue.size)
    }

    fun previous() {
        val i = queueIndex ?: return
        if (queue.isNotEmpty()) playQueueItem((i - 1 + queue.size) % queue.size)
    }

    private fun playQueueItem(index: Int) {
        if (index !in queue.indices) return
        queueIndex = index
        loadMix(queue[index])
    }

    private fun clearQueue() {
        queue.clear()
        queueIndex = null
    }

    // MARK: Transport --------------------------------------------------------

    fun togglePlayPause() {
        if (!hasSelection) return
        if (isPlaying) pause() else resume()
    }

    fun resume() {
        if (!hasSelection) return
        isPlaying = true
        requestFocus()
        selectedSounds.forEach { start(it) }
        onPlaybackBegan()
    }

    fun pause() {
        isPlaying = false
        players.forEach { (id, player) ->
            fade(id, player, target = 0f) {
                if (!isPlaying) runCatching { player.pause() }
            }
        }
        notifyChanged()
    }

    // MARK: Players ----------------------------------------------------------

    private fun start(sound: Sound) {
        val player = playerFor(sound) ?: return
        player.isLooping = true
        if (!player.isPlaying) {
            player.setVolume(0f, 0f)
            runCatching { player.start() }
        }
        fade(sound.id, player, target = level(sound.id))
    }

    private fun fadeOutAndStop(id: String) {
        val player = players[id] ?: return
        fade(id, player, target = 0f) {
            runCatching {
                player.stop()
                player.release()
            }
            players.remove(id)
        }
    }

    private fun playerFor(sound: Sound): MediaPlayer? {
        players[sound.id]?.let { return it }
        return runCatching {
            val afd = appContext.assets.openFd(sound.assetPath)
            MediaPlayer().apply {
                setAudioAttributes(audioAttributes)
                setDataSource(afd.fileDescriptor, afd.startOffset, afd.length)
                afd.close()
                isLooping = true
                prepare()
            }
        }.getOrNull()?.also { players[sound.id] = it }
    }

    private fun level(id: String): Float = (volumes[id] ?: DEFAULT_VOLUME) * masterVolume

    private fun setLevel(player: MediaPlayer, id: String) {
        val l = level(id)
        runCatching { player.setVolume(l, l) }
    }

    private fun fade(id: String, player: MediaPlayer, target: Float, onDone: (() -> Unit)? = null) {
        fadeJobs.remove(id)?.cancel()
        fadeJobs[id] = scope.launch {
            val steps = 12
            val start = currentVolume[id] ?: target
            for (i in 1..steps) {
                val v = start + (target - start) * (i / steps.toFloat())
                runCatching { player.setVolume(v, v) }
                currentVolume[id] = v
                delay(FADE_MS / steps)
            }
            currentVolume[id] = target
            onDone?.invoke()
        }
    }

    // MARK: Service + focus --------------------------------------------------

    private fun onPlaybackBegan() {
        val intent = Intent(appContext, PlaybackService::class.java)
        ContextCompat.startForegroundService(appContext, intent)
        notifyChanged()
    }

    private fun stop() {
        abandonFocus()
        notifyChanged()
        appContext.stopService(Intent(appContext, PlaybackService::class.java))
    }

    private fun notifyChanged() {
        onStateChanged?.invoke()
    }

    private fun requestFocus() {
        val request = AudioFocusRequest.Builder(AudioManager.AUDIOFOCUS_GAIN)
            .setAudioAttributes(audioAttributes)
            .setOnAudioFocusChangeListener { change ->
                when (change) {
                    AudioManager.AUDIOFOCUS_LOSS -> clearAll()
                    AudioManager.AUDIOFOCUS_LOSS_TRANSIENT,
                    AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK -> {
                        if (isPlaying) {
                            pausedForFocusLoss = true
                            pause()
                        }
                    }
                    AudioManager.AUDIOFOCUS_GAIN -> {
                        if (pausedForFocusLoss) {
                            pausedForFocusLoss = false
                            resume()
                        }
                    }
                }
            }
            .build()
        focusRequest = request
        audioManager.requestAudioFocus(request)
    }

    private fun abandonFocus() {
        focusRequest?.let { audioManager.abandonAudioFocusRequest(it) }
        focusRequest = null
        pausedForFocusLoss = false
    }

    companion object {
        private const val DEFAULT_VOLUME = 0.5f
        private const val FADE_MS = 250L
        private const val KEY_MASTER_VOLUME = "masterVolume"
    }
}
